citation-matching/citation-matching-core-code/src/main/java/pl/edu/icm/coansys/citations/CitationMatchingJob.java

```diff
@@ -0,0 +1,148 @@
+package pl.edu.icm.coansys.citations;
+
+import java.io.IOException;
+import java.util.Iterator;
+import java.util.List;
+
+import org.apache.commons.lang3.tuple.ImmutablePair;
+import org.apache.commons.lang3.tuple.Pair;
+import org.apache.hadoop.io.BytesWritable;
+import org.apache.hadoop.io.Text;
+import org.apache.hadoop.mapreduce.lib.output.SequenceFileOutputFormat;
+import org.apache.spark.SparkConf;
+import org.apache.spark.api.java.JavaPairRDD;
+import org.apache.spark.api.java.JavaSparkContext;
+
+import com.beust.jcommander.JCommander;
+import com.beust.jcommander.Parameter;
+import com.beust.jcommander.Parameters;
+import com.google.common.collect.Lists;
+
+import pl.edu.icm.coansys.citations.hashers.HashGenerator;
+import pl.edu.icm.coansys.citations.data.HashHeuristicResult;
+import pl.edu.icm.coansys.citations.data.MatchableEntity;
+import pl.edu.icm.coansys.citations.data.TextWithBytesWritable;
+
+/**
+ * 
+ * Citation matching job.
+ * 
+ * @author Łukasz Dumiszewski
+ */
+
+public class CitationMatchingJob {
+    
+    private static DocumentAttacher documentAttacher = new DocumentAttacher();
+    
+    private static CitationAttacherWithMatchedLimiter citationAttacher = new CitationAttacherWithMatchedLimiter();
+    
+    private static BestMatchedCitationPicker bestMatchedCitationPicker = new BestMatchedCitationPicker();
+    
+    
+    //------------------------ LOGIC --------------------------
+    
+    public static void main(String[] args) throws IOException, InstantiationException, IllegalAccessException, ClassNotFoundException {
+        
+        CitationMatchingJobParameters params = new CitationMatchingJobParameters();
+        JCommander jcommander = new JCommander(params);
+        jcommander.parse(args);
+        
+        
+        SparkConf conf = new SparkConf();
+        conf.set("spark.serializer", "org.apache.spark.serializer.KryoSerializer");
+        conf.registerKryoClasses(new Class[]{MatchableEntity.class});
+        
+        try (JavaSparkContext sc = new JavaSparkContext(conf)) {
+            
+            // read citations and documents
+            JavaPairRDD<Text, BytesWritable> citations = sc.sequenceFile(params.citationPath, Text.class, BytesWritable.class);
+//            citations = citations.cache(); // cache here makes the join (a few lines below) not working - the join gives 0 results
+            JavaPairRDD<Text, BytesWritable> documents = sc.sequenceFile(params.documentPath, Text.class, BytesWritable.class);
+            
+            
+            List<Pair<MatchableEntityHasher, MatchableEntityHasher>> entitiesHashers = createMatchableEntityHashers(params.hashGeneratorClasses);
```

Comment:
I would move the lines 63-78 to a private method named e.g. matchCitDocByHashes or something like that.